### DISCUSSION TITLE
Underscore for node value access?

### DISCUSSION THREAD

**Discussion by berwynhoyt** [anet-be/lua-yottadb#18] [Polls]
How shall we distinguish node value access from node subscript? Which does `node.value` do?
We have determined that we can use Lua's colon operator `node:data()` to distinguish methods from the `node.data` subnode (though it is slow because it creates a superfluous subnode `node.data` object and then invokes that with `self` whereupon it realizes that it has to return node's value instead).

**Comment by berwynhoyt** [comment]
See the relevant [pull-request and explanation here](https://github.com/anet-be/lua-yottadb/pull/16#issuecomment-1333330762).

**Reply by AlainDsc** [reply]
I would prefer 2 ways:

 - 'x = node._' as the value of node ; gives nil when not-existing
 - 'x = node:get()' as the perfect brother of M $get ; gives "" when not-existing, or the extra parameter value, like 'x = node.get("mydefault")'

**Reply by berwynhoyt** [reply]
Great idea to specify a default with `node:get("default")`. But I feel like I don't know enough about M to understand the `""`. Can you write more about that?

Returning nil make Lua idioms work like `if x then`, `assert(x)`, and even `for d in pairs(x:values())`. Are there similar idioms in M that make `""` and ideal return value for empty? I know about `$order()` where `""` is the starting and ending value in the collation sequence. Are there others? And will this behaviour still make sense in Lua? If there's more for me to know, please discuss.

**Reply by AlainDsc** [reply]
> Great idea to specify a default with `node:get("default")`. But I feel like I don't know enough about M to understand the `""`. Can you write more about that?
> 
> Returning nil make Lua idioms work like `if x then`, `assert(x)`, and even `for d in pairs(x:values())`. Are there similar idioms in M that make `""` and ideal return value for empty? I know about `$order()` where `""` is the starting and ending value in the collation sequence. Are there others? And will this behaviour still make sense in Lua? If there's more for me to know, please discuss.

In M, there is no such thing like 'nil'. If you access some global, like in 's x=^ZAD("a")', and the node does not exist, then you get an error. If you want something anyway, then you write 's x=$g(^ZAD("a"))' . In that case, you don't get an error, but a nullstring if node does not exist. There is also a variant with a second argument : 's x=$g(^ZAD("a"),"mydefault")'. In practice, the M developer will decide - case by case - if she or he prefers to let the software crash or to obfuscate the non-existence.

**Reply by berwynhoyt** [reply]
Thanks. In Lua, errors like this are typically returned as (nil, error_message). So if you want a default on empty, you'd do node:get('mydefault'). But if you want an error you'd do assert(node:get()) and since `assert()` takes two parameters (test, message), it will do nothing on success and will produce the error message which is the 2nd result from :get().

**Reply by berwynhoyt** [reply]
Tentative conclusion on a default for get(), pasted here per my offline discussion with Alain:
- Make `x = node:get()` return nil so that you can optionally raise errors with `x = assert(node:get())` or specify a default with `x = node:get("")` or even something human-readable `x = node:get("No data")`. No need to implement `node:value()`.
